File: pb/src/main/java/org/sqlbuilder/pb/joins/PbRole_VnRole.java
package org.sqlbuilder.pb.joins;

import org.sqlbuilder.common.Insertable;
import org.sqlbuilder.common.RequiresIdFrom;
import org.sqlbuilder.common.Resolvable;
import org.sqlbuilder.pb.foreign.VnRole;
import org.sqlbuilder.pb.objects.Role;
import org.sqlbuilder.pb.objects.RoleSet;

import java.util.Comparator;
import java.util.HashSet;
import java.util.Objects;
import java.util.Set;

public class PbRole_VnRole implements Insertable, Resolvable<String,Integer>
{
	static public final Comparator<PbRole_VnRole> COMPARATOR = Comparator.comparing(PbRole_VnRole::getRole).thenComparing(PbRole_VnRole::getVnRole);

	public static Set<PbRole_VnRole> SET = new HashSet<>();

	private final Role role;

	private final VnRole vnRole;

	// C O N S T R U C T O R

	public static PbRole_VnRole make(final Role role, final VnRole vnRole)
	{
		var m = new PbRole_VnRole(role, vnRole);
		SET.add(m);
		return m;
	}

	private PbRole_VnRole(final Role role, final VnRole vnRole)
	{
		this.role = role;
		this.vnRole = vnRole;
	}

	// A C C E S S

	public Role getRole()
	{
		return role;
	}

	public VnRole getVnRole()
	{
		return vnRole;
	}

	// I D E N T I T Y

	@Override
	public boolean equals(final Object o)
	{
		if (this == o)
		{
			return true;
		}
		if (o == null || getClass() != o.getClass())
		{
			return false;
		}
		PbRole_VnRole that = (PbRole_VnRole) o;
		return role.equals(that.role) && vnRole.equals(that.vnRole);
	}

	@Override
	public int hashCode()
	{
		return Objects.hash(role, vnRole);
	}

	// I N S E R T

	@RequiresIdFrom(type = Role.class)
	@RequiresIdFrom(type = RoleSet.class)
	@Override
	public String dataRow()
	{
		// rolesetid,roleid,vnclassid,vnroleid,vnclass,vntheta
		return String.format("%d,%d,%s,%s,'%s','%s'", //
				role.getRoleSet().getIntId(), //
				role.getIntId(), //
				"NULL", //
				"NULL", //
				vnRole.getVnClass().getClassName(), //
				vnRole.getVnTheta());
	}

	@Override
	public String comment()
	{
		return String.format("%s,%s", //
				role.getRoleSet().getName(), //
				role.getArgn());
	}

	// R E S O L V E

	@Override
	public String resolving()
	{
		return null;
	}

	// T O S T R I N G

	@Override
	public String toString()
	{
		return String.format("%s > %s", role, vnRole);
	}
}

File: sl/src/main/java/org/sqlbuilder/sl/foreign/PbRole_VnRole.kt
package org.sqlbuilder.sl.foreign

import org.sqlbuilder.common.Insertable
import org.sqlbuilder.common.Resolvable
import org.sqlbuilder.common.Utils.nullableInt
import org.sqlbuilder.sl.PbVnRoleResolvable
import org.sqlbuilder.sl.PbVnRoleResolved
import org.sqlbuilder.sl.SlPbRoleResolvable
import org.sqlbuilder.sl.SlVnRoleResolvable
import java.util.*

class PbRole_VnRole private constructor(
    val pbRole: PbRole,
    val vnRole: VnRole,
) : Insertable, Resolvable<PbVnRoleResolvable, PbVnRoleResolved> {

    // I D E N T I T Y

    override fun equals(o: Any?): Boolean {
        if (this === o) {
            return true
        }
        if (o == null || javaClass != o.javaClass) {
            return false
        }
        val that = o as PbRole_VnRole
        return pbRole == that.pbRole && vnRole == that.vnRole
    }

    override fun hashCode(): Int {
        return Objects.hash(pbRole, vnRole)
    }

    // I N S E R T

    override fun dataRow(): String {
        return "'${pbRole.roleSet}','${pbRole.arg}','${vnRole.vnClass}','${vnRole.theta.theta}'"
    }

    // R E S O L V E

    override fun resolving(): PbVnRoleResolvable {
        return PbVnRoleResolvable(SlPbRoleResolvable(pbRole.roleSet.lowercase(), pbRole.arg), SlVnRoleResolvable(vnRole.vnClass, vnRole.theta.theta))
    }

    // T O S T R I N G

    override fun toString(): String {
        return "$pbRole - $vnRole"
    }

    companion object {

        val COMPARATOR: Comparator<PbRole_VnRole> = Comparator
            .comparing<PbRole_VnRole, PbRole> { it.pbRole }
            .thenComparing<VnRole> { it.vnRole }

        val SET: MutableSet<PbRole_VnRole> = TreeSet<PbRole_VnRole>(COMPARATOR)

        val RESOLVE_RESULT_STRINGIFIER = { r: PbVnRoleResolved? ->
            if (r == null)
                "NULL,NULL,NULL,NULL"
            else {
                val s1 = "${nullableInt(r.first.first)},${nullableInt(r.first.second)}"
                val s2 = "${nullableInt(r.second.first)},${nullableInt(r.second.second)},${nullableInt(r.second.third)}"
                "$s1,$s2"
            }
        }

        fun make(role: PbRole, vnRole: VnRole): PbRole_VnRole {
            val a = PbRole_VnRole(role, vnRole)
            SET.add(a)
            return a
        }
    }
}
